devito/types/basic.py

```diff
@@ -1563,6 +1373,225 @@ def __getnewargs_ex__(self):
+                # doesn't work as it expects a flattened.
+                newobj = super()._new(args[2])
+
+            # Filter grid and dimensions
+            grid, dimensions = newobj._infer_dims()
+            if grid is None and dimensions is None:
+                return sympy.ImmutableDenseMatrix(*args)
+            # Initialized with constructed object
+            newobj.__init_finalize__(newobj.rows, newobj.cols, newobj.flat(),
+                                     grid=grid, dimensions=dimensions)
+        else:
+            # Initialize components and create new Matrix from standard
+            # Devito inputs
+            comps = cls.__subfunc_setup__(*args, **kwargs)
+            newobj = super()._new(comps)
+            newobj.__init_finalize__(*args, **kwargs)
+
+        return newobj
+
+    @classmethod
+    def _fromrep(cls, rep):
+        """
+        This the new constructor mechanism for matrices in sympy 1.9.
+        Standard new object go through `_new` but arithmetic operations directly use
+        the representation based one.
+        This class method is only accessible from an existing AbstractTensor
+        that contains a grid or dimensions.
+        """
+        newobj = super()._fromrep(rep)
+        grid, dimensions = newobj._infer_dims()
+        try:
+            # This is needed when `_fromrep` is called directly in 1.9
+            # for example with mul.
+            newobj.__init_finalize__(newobj.rows, newobj.cols, newobj.flat(),
+                                     grid=grid, dimensions=dimensions)
+        except TypeError:
+            # We can end up here when `_fromrep` is called through the default _new
+            # when input `comps` don't have grid or dimensions. For example
+            # `test_non_devito_tens` in `test_tensor.py`.
+            pass
+        return newobj
+
+    @classmethod
+    def __subfunc_setup__(cls, *args, **kwargs):
+        """Setup each component of the tensor as a Devito type."""
+        return []
+
+    @property
+    def grid(self):
+        """
+        A Tensor is expected to have all its components defined over the same grid
+        """
+        grids = {getattr(c, 'grid', None) for c in self.flat()} - {None}
+        if len(grids) == 0:
+            return None
+        assert len(grids) == 1
+        return grids.pop()
+
+    @property
+    def name(self):
+        for c in self.values():
+            try:
+                return c.name.split('_')[0]
+            except AttributeError:
+                # `c` is not a devito object
+                pass
+        # If we end up here, then we have no devito objects
+        # in the matrix, so we ust return the class name
+        return self.__class__.__name__
+
+    def _rebuild(self, *args, **kwargs):
+        # Plain `func` call (row, col, comps)
+        if not kwargs.keys() & self.__rkwargs__:
+            assert len(args) == 3
```

Comment:
instead of an assert I would use a ValueError, since this could easily be a user or developer programming attempted thing no? not entirely sure honetly, but in theory one can do `mytensor._rebuild(a)` where `a` is whatever. And that imho deserves a proper exception